File: Python/290.py
class Solution(object):
    def wordPattern(self, pattern, s):
        pattern = list(pattern)
        s = s.split()
        d = {}
        if len(s)!=len(pattern):
            return False
        for i in range(len(pattern)):
            if pattern[i] not in d:
                d[pattern[i]] = s[i]
            else:
                if d[pattern[i]] != s[i]:
                    return False
        d = {}
        for i in range(len(pattern)):
            if s[i] not in d:
                d[s[i]] = pattern[i]
            else:
                if d[s[i]] != pattern[i]:
                    return False
        return True

File: Java/290.java
import java.util.*;
class Solution {
    public boolean wordPattern(String pattern, String s) {
        HashMap<Character,String> d = new HashMap<>();
        ArrayList<Character> p = new ArrayList<>();
        ArrayList<String> lst = new ArrayList<>();
        for(int i=0;i<pattern.length();i++){
            p.add(pattern.charAt(i));
        }
        int start = 0;
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)==' '){
                lst.add(s.substring(start,i));
                start = i+1;
            }
        }
        lst.add(s.substring(start));
        if(p.size() != lst.size()){
            return false;
        }
        for(int i=0;i<lst.size();i++){
            if(!d.containsKey(p.get(i))){
                d.put(p.get(i),lst.get(i));
            }
            else{
                if(!(d.get(p.get(i))).equals(lst.get(i))){
                    return false;
                }
            }
        }
        HashMap<String,Character> m = new HashMap<>();
        for(int i=0;i<lst.size();i++){
            if(!m.containsKey(lst.get(i))){
                m.put(lst.get(i),p.get(i));
            }
            else{
                if(!(m.get(lst.get(i))).equals(p.get(i))){
                    return false;
                }
            }
        }
        return true;
    }
}